Source/Core/VideoCommon/VertexLoaderManager.cpp

```diff
@@ -198,57 +197,45 @@ NativeVertexFormat* GetUberVertexFormat(const PortableVertexDeclaration& decl)
   return GetOrCreateMatchingFormat(new_decl);
 }
 
-static VertexLoaderBase* RefreshLoader(int vtx_attr_group, bool preprocess = false)
+template <bool IsPreprocess>
+void CreateLoader(int vtx_attr_group)
 {
-  CPState* state = preprocess ? &g_preprocess_cp_state : &g_main_cp_state;
-  BitSet8& attr_dirty = preprocess ? g_preprocess_vat_dirty : g_main_vat_dirty;
-  auto& vertex_loaders = preprocess ? g_main_vertex_loaders : g_preprocess_vertex_loaders;
-  g_current_vat = vtx_attr_group;
+  CPState* state = IsPreprocess ? &g_preprocess_cp_state : &g_main_cp_state;
+  BitSet8& attr_dirty = IsPreprocess ? g_preprocess_vat_dirty : g_main_vat_dirty;
+  auto& vertex_loaders = IsPreprocess ? g_preprocess_vertex_loaders : g_main_vertex_loaders;
 
   VertexLoaderBase* loader;
-  if (attr_dirty[vtx_attr_group])
+
+  // We are not allowed to create a native vertex format on preprocessing as this is on the wrong
+  // thread
+  bool check_for_native_format = !IsPreprocess;
+
+  VertexLoaderUID uid(state->vtx_desc, state->vtx_attr[vtx_attr_group]);
+  std::lock_guard<std::mutex> lk(s_vertex_loader_map_lock);
+  VertexLoaderMap::iterator iter = s_vertex_loader_map.find(uid);
```

Comment:
Feels like you could avoid some of your lookups in the failed case if you use a hint:

```
auto iter = s_vertex_loader_map.lower_bound(uid);
if (iter != s_vertex_loader_map.end() && iter->first == uid)
{
  check_for_native_format &= !loader->m_native_vertex_format;
}
else
{
  iter = s_vertex_loader_map.insert(iter, VertexLoaderBase::CreateVertexLoader(state->vtx_desc, state->vtx_attr[vtx_attr_group]));
  INCSTAT(g_stats.num_vertex_loaders);
}
loader = iter->second.get();
```

Reply:
Unordered maps don't have a lower bound function, only regular (ordered) maps do.
We could probably do `std::lower_bound`. But that would end up doing a linear search, wouldn't it?
I guess we could switch the VertexLoaderMap to an ordered map but I'm not sure if the hint would make up for the worse lookup performance (O(log(n)) instead of O(1)).

Either way, it seems that this isn't particularly hot code anyway, so I'll probably just leave it as it is. 90%+ of calls here are handled by the dirty flag and just go through the simple array.

Reply:
I missed that this was an unordered map.  Even if this isn't a "hot" path, I'd at least change the failed case where you are doing *two* lookups below:

```
auto [it, added] = s_vertex_loader_map.try_emplace(uid, VertexLoaderBase::CreateVertexLoader(state->vtx_desc, state->vtx_attr[vtx_attr_group]));
loader = it->second.get();
INCSTAT(g_stats.num_vertex_loaders);
```

Reply:
Done.